File: mac/Tests/BG3AssistantTests/OnboardingTests.swift
import XCTest
@testable import BG3HonorAssistant

final class OnboardingTests: XCTestCase {
    private func step(id: String, order: Int, checkpointId: String? = nil) -> WalkthroughStep {
        WalkthroughStep(
            id: id, order: order, phase: "Phase", phaseOrder: 1,
            title: "Title \(id)", kind: "exploration", importance: "core", region: "Wilderness",
            area: "Area", minimumLevel: 1, summary: "", avoid: "", why: "",
            rewards: [], completionChecks: [], prerequisites: [], dependencies: [],
            checkpointId: checkpointId, markerId: nil, decision: nil, incident: nil,
            riskReward: nil, authority: "guide_fact", sourceLabel: "", sourceUrl: ""
        )
    }

    func testProviderSetupIsRequiredForBothOnboardingPaths() {
        XCTAssertEqual(OnboardingStep.steps(for: .fresh), [.welcome, .ai, .party, .ready])
        XCTAssertEqual(OnboardingStep.steps(for: .midRun), [.welcome, .ai, .party, .catchUp, .ready])
    }

    func testLocalModelCapabilitiesAreExplicit() {
        XCTAssertEqual(AIProvider.localGemma.ollamaModel, "gemma4:12b")
        XCTAssertEqual(AIProvider.localGemma.modelDownloadSize, "7.6 GB")
        XCTAssertTrue(AIProvider.localGemma.supportsImages)

        XCTAssertEqual(AIProvider.localQwen.ollamaModel, "qwen3:4b")
        XCTAssertEqual(AIProvider.localQwen.modelDownloadSize, "2.5 GB")
        XCTAssertFalse(AIProvider.localQwen.supportsImages)
    }

    func testOllamaImageIsAttachedToLastMessage() throws {
        let encoded = AssistantAIClient.ollamaMessages(
            [
                AssistantAIMessage(role: "system", content: "Use the guide."),
                AssistantAIMessage(role: "user", content: "What is on screen?"),
            ],
            imageData: Data([1, 2, 3])
        )

        XCTAssertNil(encoded[0]["images"])
        XCTAssertEqual(try XCTUnwrap(encoded[1]["images"] as? [String]), ["AQID"])
    }

    func testCatchUpMarksThroughLandmarkInclusive() throws {
        let walkthrough = [
            step(id: "s1", order: 1, checkpointId: "cp1"),
            step(id: "s2", order: 2),
            step(id: "s3", order: 3, checkpointId: "cp2"),
            step(id: "s4", order: 4),
        ]
        let ledger = try XCTUnwrap(CatchUp.ledger(markingThrough: "cp2", walkthrough: walkthrough, existing: [:]))
        XCTAssertEqual(ledger["s1"], .caughtUp)
        XCTAssertEqual(ledger["s2"], .caughtUp)
        XCTAssertEqual(ledger["s3"], .caughtUp)
        XCTAssertNil(ledger["s4"])
    }

    func testCatchUpPreservesExplicitHistory() throws {
        let walkthrough = [
            step(id: "s1", order: 1, checkpointId: "cp1"),
            step(id: "s2", order: 2, checkpointId: "cp2"),
        ]
        let ledger = try XCTUnwrap(CatchUp.ledger(
            markingThrough: "cp2", walkthrough: walkthrough,
            existing: ["s1": .skipped]
        ))
        XCTAssertEqual(ledger["s1"], .skipped)
        XCTAssertEqual(ledger["s2"], .caughtUp)
    }
}

File: windows/tests/BG3HonorAssistant.Core.Tests/Models/OnboardingTests.cs
using BG3HonorAssistant.Core.Models;

namespace BG3HonorAssistant.Core.Tests.Models;

public sealed class OnboardingTests
{
    [Fact]
    public void ProviderSetupStepIsPresentForBothOnboardingPaths()
    {
        Assert.Equal(
            [
                OnboardingStep.Welcome,
                OnboardingStep.Difficulty,
                OnboardingStep.Spoilers,
                OnboardingStep.Ai,
                OnboardingStep.Party,
                OnboardingStep.Ready,
            ],
            OnboardingFlow.Steps(OnboardingMode.Fresh));
        Assert.Equal(
            [
                OnboardingStep.Welcome,
                OnboardingStep.Difficulty,
                OnboardingStep.Spoilers,
                OnboardingStep.Ai,
                OnboardingStep.Party,
                OnboardingStep.CatchUp,
                OnboardingStep.Ready,
            ],
            OnboardingFlow.Steps(OnboardingMode.MidRun));
    }

    [Fact]
    public void OnlyKnownOverlayDifficultiesAreSelectable()
    {
        Assert.Equal(
            [
                RunDifficulty.Balanced,
                RunDifficulty.Tactician,
                RunDifficulty.Honour,
            ],
            RunDifficultyExtensions.SelectableOverlayDifficulties);
        Assert.False(RunDifficulty.Explorer.SupportsOverlay());
        Assert.True(RunDifficulty.Custom.SupportsOverlay());
        Assert.False(
            OnboardingFlow.AllowsAdvance(
                OnboardingStep.Difficulty,
                RunDifficulty.Explorer));
        Assert.True(
            OnboardingFlow.AllowsAdvance(
                OnboardingStep.Difficulty,
                RunDifficulty.Balanced));
        Assert.True(
            OnboardingFlow.AllowsAdvance(
                OnboardingStep.Spoilers,
                RunDifficulty.Explorer));
    }

    [Fact]
    public void WindowsMvpExplicitlyExcludesLocalModels()
    {
        Assert.False(WindowsMvpCapabilities.SupportsLocalAi);
    }

    [Fact]
    public void WindowsMvpExplicitlyExcludesImageMessages()
    {
        Assert.False(WindowsMvpCapabilities.SupportsImageMessages);
        Assert.False(WindowsMvpCapabilities.SupportsScreenshotCapture);
        Assert.False(WindowsMvpCapabilities.SupportsMicrophoneOrSpeech);
    }

    [Fact]
    public void CatchUpMarksThroughLandmarkInclusive()
    {
        var walkthrough = new[]
        {
            Step("s1", 1, "cp1"),
            Step("s2", 2),
            Step("s3", 3, "cp2"),
            Step("s4", 4),
        };

        var ledger = CatchUp.Ledger(
            "cp2",
            walkthrough,
            new Dictionary<string, CheckpointDisposition>());

        Assert.NotNull(ledger);
        Assert.Equal(CheckpointDisposition.CaughtUp, ledger["s1"]);
        Assert.Equal(CheckpointDisposition.CaughtUp, ledger["s2"]);
        Assert.Equal(CheckpointDisposition.CaughtUp, ledger["s3"]);
        Assert.False(ledger.ContainsKey("s4"));
    }

    [Fact]
    public void CatchUpPreservesExplicitHistory()
    {
        var walkthrough = new[]
        {
            Step("s1", 1, "cp1"),
            Step("s2", 2, "cp2"),
        };

        var ledger = CatchUp.Ledger(
            "cp2",
            walkthrough,
            new Dictionary<string, CheckpointDisposition>
            {
                ["s1"] = CheckpointDisposition.Skipped,
            });

        Assert.NotNull(ledger);
        Assert.Equal(CheckpointDisposition.Skipped, ledger["s1"]);
        Assert.Equal(CheckpointDisposition.CaughtUp, ledger["s2"]);
    }

    private static WalkthroughStep Step(
        string id,
        int order,
        string? checkpointId = null) =>
        new()
        {
            Id = id,
            Order = order,
            Phase = "Phase",
            PhaseOrder = 1,
            Title = $"Title {id}",
            Kind = "exploration",
            Importance = "core",
            Region = "Wilderness",
            Area = "Area",
            MinimumLevel = 1,
            CheckpointId = checkpointId,
        };
}
